src/core/geometry/qgsellipse.h

```diff
@@ -0,0 +1,219 @@
+    virtual bool operator !=( const QgsEllipse &elp ) const;
+
+    //! An ellipse is empty if axes are equal to 0
+    virtual bool isEmpty() const;
+
+    /** Returns the center point.
+     * @see setCenter()
+     * @see rcenter()
+     */
+    QgsPointV2 center() const {return mCenter; }
+
+    /** Returns the semi-major axis.
+     * @see setSemiMajorAxis()
+     */
+    double semiMajorAxis() const {return mSemiMajorAxis; }
+
+    /** Returns the semi-minor axis.
+     * @see setSemiMinorAxis()
+     */
+    double semiMinorAxis() const {return mSemiMinorAxis; }
+
+    /** Returns the azimuth.
+     * @see setAzimuth()
+     */
+    double azimuth() const {return mAzimuth; }
+
+    /** Returns a reference to the center point of this ellipse.
+     * Using a reference makes it possible to directly manipulate center in place.
+     * @see center()
+     * @see setCenter()
+     * @note not available in Python bindings
+     */
+    QgsPointV2 &rcenter() { return mCenter; }
+
+    /** Sets the center point.
+     * @see center()
+     * @see rcenter()
+     */
+    void setCenter( const QgsPointV2 &center ) {mCenter = center; }
+
+    /** Sets the semi-major axis.
+     * @see semiMajorAxis()
+     */
+    virtual void setSemiMajorAxis( const double semiMajorAxis );
+
+    /** Sets the semi-minor axis.
+     * @see semiMinorAxis()
+     */
+    virtual void setSemiMinorAxis( const double semiMinorAxis );
+
+    /** Sets the azimuth (orientation).
+     * @see azimuth()
+     */
+    void setAzimuth( const double azimuth );
+
+    /** The distance between the center and each foci.
+     * @see fromFoci()
+     * @see foci()
+     * @return The distance between the center and each foci.
+     */
+    virtual double focusDistance() const;
+
+    /** Two foci of the ellipse. The axes are oriented by the azimuth and are on the semi-major axis.
+     * @see fromFoci()
+     * @see focusDistance()
+     * @return the two foci.
+     */
+    virtual QVector<QgsPointV2> foci() const;
+
+    //! The eccentricity of the ellipse.
+    virtual double eccentricity() const;
+    //! The area of the ellipse.
+    virtual double area() const;
+    //! The circumference of the ellipse using first approximation of Ramanujan.
+    virtual double perimeter() const;
+
+    /** The four quadrant of the ellipse.
+     * The are oriented and started always from semi-major axis.
+     * @return quadrant defined by four point.
+     */
+    virtual QVector<QgsPointV2> quadrant() const;
+
+    /** Returns a list of points into \a pts, with segmentation from \a segments.
+     * @param pts List of points returned.
+     * @param segments Number of segments used to segment geometry.
+     */
+    virtual void points( QgsPointSequence &pts, unsigned int segments = 36 ) const;
+
+    /** Returns a segmented polygon.
+     * @param segments Number of segments used to segment geometry.
+     */
+    virtual QgsPolygonV2 toPolygon( unsigned int segments = 36 ) const;
```

Comment:
I think given that the main use for toPolygon and toLineString will be store the result in a QgsGeometry that these should be factories and return a pointer. It'll avoid the need for an extra clone to store the result in QgsGeometry.